trust-quorum/src/proxy.rs

```diff
@@ -0,0 +1,589 @@
+    ops: IdHashMap<TrackableRequest>,
+}
+
+impl Tracker {
+    pub fn new() -> Tracker {
+        Tracker { ops: IdHashMap::new() }
+    }
+
+    /// The number of proxied requests outstanding
+    pub fn len(&self) -> usize {
+        self.ops.len()
+    }
+
+    pub fn insert(&mut self, req: TrackableRequest) {
+        self.ops.insert_unique(req).expect("no duplicate request IDs");
+    }
+
+    /// Handle a proxied response from a peer
+    pub fn on_response(
+        &mut self,
+        request_id: Uuid,
+        result: Result<WireValue, WireError>,
+    ) {
+        if let Some(mut req) = self.ops.remove(&request_id) {
+            let res = result.map_err(TrackerError::Wire);
+            let _ = req.tx.take().unwrap().send(res);
+        }
+    }
+
+    /// A remote peer has disconnected
+    pub fn on_disconnect(&mut self, from: &BaseboardId) {
+        self.ops.retain(|mut req| {
+            if &req.destination == from {
+                let tx = req.tx.take().unwrap();
+                let _ = tx.send(Err(TrackerError::Disconnected));
+                false
+            } else {
+                true
+            }
+        });
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use assert_matches::assert_matches;
+    use omicron_test_utils::dev::poll::{CondCheckError, wait_for_condition};
+    use std::sync::{
+        Arc,
+        atomic::{AtomicUsize, Ordering},
+    };
+    use std::time::Duration;
+    use tokio::spawn;
+
+    /// Recv a message from the proxy and insert it into the tracker.
+    async fn recv_and_insert(
+        rx: &mut mpsc::Receiver<NodeApiRequest>,
+        tracker: &mut Tracker,
+    ) {
+        let Some(NodeApiRequest::Proxy { destination, wire_request, tx }) =
+            rx.recv().await
+        else {
+            panic!("Invalid NodeApiRequest")
+        };
+
+        let req =
+            TrackableRequest::new(destination, wire_request.request_id, tx);
+        tracker.insert(req);
+    }
+
+    #[tokio::test]
+    async fn proxy_roundtrip_concurrent() {
+        let destination = BaseboardId {
+            part_number: "test".to_string(),
+            serial_number: "test".to_string(),
+        };
+        let rack_id = RackUuid::new_v4();
+
+        // Test channel where the sender is usually cloned from the [`crate::NodeTaskHandle`],
+        // and the receiver is owned by the local [`crate::NodeTask`]
+        let (tx, mut rx) = mpsc::channel(5);
+        let proxy = Proxy::new(tx.clone());
+        let mut tracker = Tracker::new();
+
+        let requests_completed = Arc::new(AtomicUsize::new(0));
+
+        // This is the first "user" task that will issue proxy operations
+        let count = requests_completed.clone();
+        let dest = destination.clone();
+        let _ = spawn(async move {
+            let s = proxy.commit(dest, rack_id, Epoch(1)).await.unwrap();
+
+            // The first attempt should succeed
+            assert_eq!(s, CommitStatus::Committed);
```

Comment:
We are not joining this task at any point, is there a risk that this assert doesn't get executed?

Reply:
Comment added above `let requests_completed` in [e9774cf](https://github.com/oxidecomputer/omicron/pull/9403/commits/e9774cf9b2b0279f3569c1523062a9b02fa51508)